### PR TITLE
add run command code base

### PR DESCRIPTION
**Pull request by calvix**
one of small Prs built from https://github.com/giantswarm/etcd-cluster-migrator/pull/1

the migration expects that cluster was created as a single node cluster and then it was upgraded to 3 nodes. So 1 etcd is already bootstrapped and the other two are not.

The whole process of adding node consist of 2 phases:
1 -  changing etcd3.service on the node so the initial cluster state is set for a node to join the cluster and then restarting the etcd3 service. This is done via pod which executes commands on the node.
2 - adding a new node to the etcdcluster database via etcd API

After that, we wait until the etcd node sync data and cluster is available again

This process is the same for the second and third node. 

run command functionality creates a job on a specific node and executed command in the host namespace ( like executing from the machine itself. The commands are saved into the CM and mounted as volume do the pod.

Circle CI is red due code spread into multiple PRs, but as you can on the original big PR its green.

Other related PRs:
base components  - https://github.com/giantswarm/etcd-cluster-migrator/pull/2
migration code - https://github.com/giantswarm/etcd-cluster-migrator/pull/4

### CODE REVIEW

File: helm/etcd-cluster-migrator/templates/job-psp-clusterrole.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (xh3b4sd) [inline helm/etcd-cluster-migrator/templates/job-psp-clusterrole.yaml]: This is `v1` in the `aws-operator`. Any reason to go with `v1beta1`?

File: helm/etcd-cluster-migrator/templates/job-psp.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+apiVersion: policy/v1beta1
+kind: PodSecurityPolicy
+metadata:
+  name: {{ .Values.name }}-job
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app: {{ .Values.name }}-job
+    giantswarm.io/service-type: "managed"
```
Comment (xh3b4sd) [inline helm/etcd-cluster-migrator/templates/job-psp.yaml]: We use `giantswarm.io/managed-by` when it is managed by us. See also https://github.com/giantswarm/fmt/blob/master/kubernetes/annotations_and_labels.md#common-labels.

File: helm/etcd-cluster-migrator/templates/job-serviceaccount.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ .Values.name }}-job
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app: {{ .Values.name }}-job
+    giantswarm.io/service-type: "managed"
```
Comment (xh3b4sd) [inline helm/etcd-cluster-migrator/templates/job-serviceaccount.yaml]: Newline would be grand.

File: migrator/cmd.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+			job, err := m.k8sClient.BatchV1().Jobs(runCommandNamespace).Get(job.Name, apismetav1.GetOptions{})
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			if isJobCompleted(job) {
+				err := m.k8sClient.BatchV1().Jobs(runCommandNamespace).Delete(job.Name, &apismetav1.DeleteOptions{})
+
+				fmt.Printf("Job %s was completed.\n", job.Name)
```
Comment (xh3b4sd) [inline migrator/cmd.go]:
```suggestion
				fmt.Printf("Job %s was completed.\n", job.Name)

				err := m.k8sClient.BatchV1().Jobs(runCommandNamespace).Delete(job.Name, &apismetav1.DeleteOptions{})
```

File: migrator/cmd.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+						{
+							Key:      "node.kubernetes.io/unschedulable",
+							Operator: "Exists",
+							Effect:   "NoSchedule",
+						},
+						{
+							Key:    "node-role.kubernetes.io/master",
+							Effect: "NoSchedule",
+						},
+					},
```
Comment (xh3b4sd) [inline migrator/cmd.go:177]: FMI, I can never really read and understand what the tolerations mean. Does that mean that the job is tolerating the "master no schedule" taint or that it cannot be scheduled on nodes defining this toleration? 🙈

**Review by xh3b4sd — APPROVED** [review]
I would appreciate feedback from somebody like @corest and @whites11 maybe as well. I personally cannot find anything particularly wrong with it besides style. When this works let's go.

File: helm/etcd-cluster-migrator/templates/job-psp-clusterrole.yaml (hunk 1 above)
Comment (calvix) [reply helm/etcd-cluster-migrator/templates/job-psp-clusterrole.yaml]: v1 is better, I copypasted from kiam-agent where for some reason we have the v1beta1

I will fix

File: helm/etcd-cluster-migrator/templates/job-psp.yaml (hunk 2 above)
Comment (calvix) [reply helm/etcd-cluster-migrator/templates/job-psp.yaml]: ack, adding `giantswarm.io/managed-by: "aws-operator"` for now

File: helm/etcd-cluster-migrator/templates/job-serviceaccount.yaml (hunk 3 above)
Comment (calvix) [reply helm/etcd-cluster-migrator/templates/job-serviceaccount.yaml]: fixed

File: migrator/cmd.go (hunk 5 above)
Comment (calvix) [reply migrator/cmd.go:177]: it basically means, ignore this taint 

so schedule even if this taint is on that node

**Comment by whites11** [comment]
> 1 - changing etcd3.service on the node so the initial cluster state is set for a node to join the cluster and then restarting the etcd3 service. This is done via pod which executes commands on the node.

Why do you even need this? the --initial-cluster-state flag is ignored if the etcd node is already bootstrapped. You can remove it AFAIK. You just need to `etcdctl member add` the other two nodes and you're good to go. Am I wrong?

**Review by whites11 — APPROVED** [review]
LGTM

**Comment by calvix** [comment]
> Why do you even need this? the --initial-cluster-state flag is ignored if the etcd node is already bootstrapped. You can remove it AFAIK. You just need to etcdctl member add the other two nodes and you're good to go. Am I wrong?

node is not bootstrapped until it joins a cluster so you need to set proper initial cluster state

**Pull request merged**